src/bit-fields-macros/src/builder.rs

```diff
@@ -0,0 +1,816 @@
+
+        let header = format!(
+            "A {} bit structure containing a number of bit flags and bit fields.",
+            self.data_type.size()
+        );
+
+        quote! {
+            #[doc=#header]
+            ///
+            /// ## Layout
+            ///
+            /// <table>
+            #struct_doc_table_layout
+            /// </table>
+            #[derive(Debug,Clone,Copy,PartialEq,Eq)]
+            #[repr(C)]
+            pub struct #struct_name(pub #struct_data_type);
+            #serde
+            impl std::cmp::PartialEq<#struct_data_type> for #struct_name {
+                fn eq(&self, other: &#struct_data_type) -> bool {
+                    self.0 == *other
+                }
+            }
+            #display_impl
+            /// We cannot derive [`std::default::Default`] as `self.bits` may have too many elements.
+            impl std::default::Default for #struct_name {
+                fn default() -> Self {
+                    Self::from(0)
+                }
+            }
+            impl std::fmt::Binary for #struct_name {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                    std::fmt::Binary::fmt(&self.0, f)
+                }
+            }
+            impl std::fmt::LowerHex for #struct_name {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                    std::fmt::LowerHex::fmt(&self.0, f)
+                }
+            }
+            impl std::fmt::Octal for #struct_name {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                    std::fmt::Octal::fmt(&self.0, f)
+                }
+            }
+            #flag_set
+            #field_map
+            #set_map_conversions
+            /// Constructs `self` from the data type.
+            impl std::convert::From<#struct_data_type> for #struct_name {
+                fn from(data: #struct_data_type) -> Self {
+                    Self(data)
+                }
+            }
+            /// Constructs the data type from `self`.
+            impl std::convert::From<#struct_name> for #struct_data_type {
+                fn from(bit_field: #struct_name) -> Self {
+                    bit_field.0
+                }
+            }
+            impl std::ops::BitOr for #struct_name {
+                type Output = Self;
+                fn bitor(self,rhs: Self) -> Self::Output {
+                    Self::from(self.0 | rhs.0)
+                }
+            }
+            impl std::ops::BitAnd for #struct_name {
+                type Output = Self;
+                fn bitand(self,rhs: Self) -> Self::Output {
+                    Self::from(self.0 & rhs.0)
+                }
+            }
+            impl std::ops::Not for #struct_name {
+                type Output = Self;
+                fn not(self) -> Self::Output {
+                    Self::from(!self.0)
```

Comment:
I think they make sense in the context of bit flags and they make other operations more convenient. I can see it is a little confusing given bit ranges but I think that this applying to the underlying raw integer is fairly logical.

Reply:
I agree they make sense in the context of bit flags. Maybe the impl should only be provided if there are no ranges, and if there are ranges the user has to explicitly do `!field.0` (similar to what I suggest above for the `Eq` implementations), or call a specific `negate_all` method? 

My main point here is that I think we should avoid potentially surprising behavior. To me, it would be just as reasonable to have `Not` only flip the `Bit`s, but leave ranges untouched. A user expecting this might be surprised by the impl also affecting ranges. 

Either way, at the very least there should be a doc comment explaining the behavior

Reply:
Added the comment:
```rust
/// Implemented operations such as [`std::cmp::Eq`], [`std::ops::BitOr`] and
/// [`std::ops::Not`] etc. apply to the underlying integer type.
```